File: patterns/10_proxy/csharp/Legatus.cs
// ============================================================
//  PROXY PATTERN — The Imperial Legatus
//  Run: dotnet script Legatus.cs
// ============================================================
#nullable enable
using System;
using System.Collections.Generic;

// ── Interface (shared by real service and proxy) ──────────────
interface IImperialService {
    string IssueDecree(string province, string decree);
    string GetCensus(string province);
}

// ── Real Subject ──────────────────────────────────────────────
class EmperorHadrian : IImperialService {
    public EmperorHadrian() {
        Console.WriteLine("  👑 Emperor Hadrian awakens from his nap!");
    }
    public string IssueDecree(string province, string decree) =>
        $"👑 DECREE for {province}: '{decree}' — Hadrianus P.P.";
    public string GetCensus(string province) =>
        $"👑 Census of {province}: ~50,000 inhabitants";
}

// ── Proxy (access control + caching + lazy init) ──────────────
class LegateProxy : IImperialService {
    private EmperorHadrian? _emperor;
    private readonly Dictionary<string, string> _cache = new Dictionary<string, string>();
    private readonly HashSet<string> _authorized;
    private int _count;

    public LegateProxy() {
        _authorized = new HashSet<string> { "Syria", "Aegyptus", "Britannia", "Gallia", "Hispania" };
    }

    private EmperorHadrian GetEmperor() {
        if (_emperor == null) {
            Console.WriteLine("  [Proxy] Waking the Emperor...");
            _emperor = new EmperorHadrian();
        }
        return _emperor;
    }

    private void Log(string action) =>
        Console.WriteLine($"  [Log #{++_count}] {action}");

    public string IssueDecree(string province, string decree) {
        Log($"Decree requested for '{province}'");
        if (!_authorized.Contains(province))
            return $"🚫 DENIED: '{province}' is not an authorized province!";
        return GetEmperor().IssueDecree(province, decree);
    }

    public string GetCensus(string province) {
        Log($"Census requested for '{province}'");
        string? cached;
        if (_cache.TryGetValue(province, out cached)) {
            Console.WriteLine($"  [Cache HIT] Returning cached census for {province}");
            return cached;
        }
        var result = GetEmperor().GetCensus(province);
        _cache[province] = result;
        return result;
    }

    public int RequestCount => _count;
}

// ── Demo ──────────────────────────────────────────────────────
Console.WriteLine("╔═══════════════════════════════════════════════╗");
Console.WriteLine("║   PROXY PATTERN — C#                          ║");
Console.WriteLine("║   Legatus Augusti                             ║");
Console.WriteLine("╚═══════════════════════════════════════════════╝\n");

var proxy = new LegateProxy();

Console.WriteLine("── ACCESS CONTROL ──────────────────────────────────");
Console.WriteLine(proxy.IssueDecree("Barbaria", "Build roads!"));
Console.WriteLine(proxy.IssueDecree("Britannia", "Build Hadrian's Wall"));

Console.WriteLine("\n── CACHING PROXY ───────────────────────────────────");
Console.WriteLine(proxy.GetCensus("Aegyptus"));
Console.WriteLine(proxy.GetCensus("Aegyptus")); // second call: cache hit

Console.WriteLine($"\n  Total requests handled: {proxy.RequestCount}");
Console.WriteLine("\"Legatus vocem imperatoris habet!\"");

File: patterns/10_proxy/java/Legatus.java
// ============================================================
//  PROXY PATTERN — The Imperial Legatus
//  Run: java Legatus.java
// ============================================================
import java.util.*;

public class Legatus {

    interface IImperialService {
        String issueDecree(String province, String decree);
        String getCensus(String province);
    }

    static class EmperorHadrian implements IImperialService {
        EmperorHadrian() { System.out.println("  👑 Emperor Hadrian awakens from his nap!"); }
        public String issueDecree(String p, String d) {
            return "👑 IMPERIAL DECREE for " + p + ": '" + d + "' — Signed, Hadrianus P.P.";
        }
        public String getCensus(String p) {
            return "👑 Census of " + p + ": ~50,000 inhabitants (all paying taxes)";
        }
    }

    static class LegateProxy implements IImperialService {
        private EmperorHadrian emperor;
        private final Map<String, String> cache = new HashMap<>();
        private final Set<String> authorized = new HashSet<>(
            Arrays.asList("Syria","Aegyptus","Britannia","Gallia","Hispania"));
        private int count = 0;

        private EmperorHadrian getEmperor() {
            if (emperor == null) {
                System.out.println("  [Proxy] First real request — waking the Emperor...");
                emperor = new EmperorHadrian();
            }
            return emperor;
        }
        private void log(String action) {
            System.out.printf("  [Log #%d] %s%n", ++count, action);
        }
        public String issueDecree(String p, String d) {
            log("Decree for '" + p + "'");
            if (!authorized.contains(p))
                return "🚫 ACCESS DENIED: '" + p + "' is not an authorized province!";
            return getEmperor().issueDecree(p, d);
        }
        public String getCensus(String p) {
            log("Census for '" + p + "'");
            if (cache.containsKey(p)) {
                System.out.println("  [Cache HIT] " + p);
                return cache.get(p);
            }
            String r = getEmperor().getCensus(p);
            cache.put(p, r);
            return r;
        }
        int getCount() { return count; }
    }

    public static void main(String[] args) {
        System.out.println("╔═══════════════════════════════════════════════╗");
        System.out.println("║   PROXY PATTERN — Java                        ║");
        System.out.println("║   The Imperial Legatus                        ║");
        System.out.println("╚═══════════════════════════════════════════════╝\n");

        var proxy = new LegateProxy();

        System.out.println("── PROTECTION: Unauthorized ────────────────────");
        System.out.println(proxy.issueDecree("Barbaria", "Build roads!"));

        System.out.println("\n── AUTHORIZED (lazy-init Emperor) ──────────────");
        System.out.println(proxy.issueDecree("Britannia", "Build Hadrian's Wall"));

        System.out.println("\n── CACHING: Census ──────────────────────────────");
        System.out.println(proxy.getCensus("Aegyptus"));
        System.out.println("\n  (Second call — cached)");
        System.out.println(proxy.getCensus("Aegyptus"));

        System.out.println("\n  Total requests: " + proxy.getCount());
        System.out.println("\"Legatus vocem imperatoris habet!\"");
    }
}
